decisions/DR007-Hierarchical-or-compositional-traits-for-specifications.md

```diff
@@ -0,0 +1,361 @@
+    _type = ".".join(sorted((BlobTrait._type, ImageTrait._type)))
+
+
+class RasterImageSpecification(BlobTrait, ImageTrait, RasterTrait):
+    """
+    Raster image, as opposed to e.g. SVG.
+    """
+    _type = ".".join(sorted((BlobTrait._type, ImageTrait._type, RasterTrait._type)))
+
+
+class ImageBundleSpecification(ContainerTrait, ImageTrait):
+    _type = ".".join(sorted((ContainerTrait._type, ImageTrait._type)))
+
+
+class SceneCacheSpecification(GeometryTrait, CacheTrait, BlobTrait):
+    _type = ".".join(sorted((GeometryTrait._type, CacheTrait._type)))
+
+
+class EditorialSpecification(ContainerTrait, EditorialTrait):
+    _type = ".".join(sorted((ContainerTrait._type, EditorialTrait._type)))
+
+
+class ShotSpecification(ContainerTrait, EditorialTrait, ShotTrait):
+    _type = ".".join(sorted((ContainerTrait._type, EditorialTrait._type, ShotTrait._type)))
+
+
+class DocumentSpecification(BlobTrait, DocumentTrait):
+    _type = ".".join(sorted((BlobTrait._type, DocumentTrait._type)))
+
+
+class NukeScriptSpecification(BlobTrait, DocumentTrait, NukeScriptTrait):
+    _type = ".".join(sorted((BlobTrait._type, DocumentTrait._type, NukeScriptTrait._type)))
+
+
+class KatanaProjectSpecification(BlobTrait, DocumentTrait, KatanaProjectTrait):
+    _type = ".".join(sorted((BlobTrait._type, DocumentTrait._type, KatanaProjectTrait._type)))
+```
+
+### An example of using these specifications
+
+#### Checking management policies
+
+```python
+# uses traits from "An implementation sketch of Option 2"
+imageSpec = ImageSpecification()
+policy = manager.managementPolicy([imageSpec], context)[0]
+
+if policy == [kIgnored]:
+    return  # nothing with an ImageSpecification is managed
+
+# check if pngs and alembic caches are supported by the manager
+exrSpec = ImageSpecification()
+exrSpec.mimeType = "image/x-exr; version=\"2\""
+
+abcSpec = SceneCacheSpecification()
+abcSpec.mimeType = "alembic"
+policy = manager.managementPolicy([exrSpec, abcSpec], context)[0]
+
+policy == [kManaged, kManaged]  # manager offers to manage these exrs and alembic
+```
+
+#### Finding all the caches associated with a shot - (such as to schedule for deletion to free up space)
+
+```python
+
+assets = ["asset://fxPointsCache",  # a geometry cache of point data
+          "asset://precompShadows"]  # rendered image of shadows
+
+shots = ["shot://ABC/01"]
+
+
+class CacheTrait(specifications.EntityTrait):
+    _type = "cache"
+
+
+class ImageTrait(specifications.EntityTrait):
+    _type = "image"
+
+
+class ImageCacheSpec(ImageTrait, CacheTrait):
+    _type = ".".join(sorted((ImageTrait._type, CacheTrait)))
+
+
+class GeometryTrait(specifications.EntityTrait):
+    _type = "geometry"
+
+
+class GeometryCacheSpec(GeometryTrait, CacheTrait):
+    _type = ".".join(sorted((GeometryTrait._type, CacheTrait)))
+
+
+class Bundle(specifications.RelationshipSpecification):
+    _type = "bundle"
+
+
+manager.getRelatedReferences(shots, Bundle, context, session,
```

Comment:
[fix] In the current API, you woudln't need the `Bundle` spec as you can query the specifications directly:

`getRelatedReferences(shots, [ImageCacheSpec(), GeometryCacheSpec()], context, session)`

(`resultSpect` is currently singular)

Reply:
How does the manager know which caches are associated with the shots?